src/VisualStudio/LiveShare/Impl/Client/RemoteLanguageServiceWorkspace.cs

```diff
@@ -97,97 +113,123 @@ void IRunningDocumentTableEventListener.OnRenameDocument(string newMoniker, stri
-                var lastSlash = _remoteRootPath.LastIndexOf('\\');
-                _externalPath = _remoteRootPath.Substring(0, lastSlash + 1);
-                _externalPath += "~external";
-            }
             IsRemoteSession = true;
+
+            // Files could have been opened before the collaboration session was fully initialized.
+            // So get the roots and update remote paths for all open files.
+            await UpdatePathsToRemoteFilesAsync().ConfigureAwait(false);
+
             session.RemoteServicesChanged += (object sender, RemoteServicesChangedEventArgs e) =>
             {
                 _remoteDiagnosticListTable.UpdateWorkspaceDiagnosticsPresent(_session.RemoteServiceNames.Contains("workspaceDiagnostics"));
             };
         }
 
-        public void EndSession()
+        public bool IsExternalLocalUri(string localPath)
+            => _registeredExternalPaths.Any(externalPath => localPath.StartsWith(externalPath) && localPath.Length > externalPath.Length + 1);
+
+        public string GetRemoteWorkspaceRoot(string filePath)
+            => _remoteWorkspaceRootPaths.SingleOrDefault(remoteWorkspaceRoot => filePath.StartsWith(remoteWorkspaceRoot));
+
+        /// <summary>
+        /// Event that gets triggered whenever the active workspace changes.  If we're in a live share session
+        /// this means that the remote workpace roots have also changed and need to be updated.
+        /// </summary>
+        private async Task OnActiveWorkspaceChangedAsync(object sender, EventArgs args)
         {
-            IsRemoteSession = false;
+            if (IsRemoteSession)
+            {
+                await UpdatePathsToRemoteFilesAsync().ConfigureAwait(false);
+            }
         }
 
-        public void Init()
+        /// <summary>
+        /// Retrieves the base folder paths for files on the client that have been retrieved from the remote host.
+        /// Triggers a refresh of all open files so we make sure they are in the correct workspace.
+        /// </summary>
+        private async Task UpdatePathsToRemoteFilesAsync()
         {
-            StartSolutionCrawler();
-        }
+            if (_session == null)
+            {
+                return;
+            }
 
-        /// <inheritdoc />
-        public override bool CanOpenDocuments => true;
+            var roots = await _session.ListRootsAsync(CancellationToken.None).ConfigureAwait(false);
+            if (roots.Length > 0)
+            {
+                var localPathsOfRemoteRoots = roots.Select(root => _session.ConvertSharedUriToLocalPath(root)).ToImmutableArray();
 
-        /// <inheritdoc />
-        public void OnManagedProjectAdded(ProjectInfo projectInfo)
-        {
-            //Notify AnyCode Roslyn workspace a project is being added.
-            OnProjectAdded(projectInfo);
-        }
+                var remoteRootPaths = new HashSet<string>();
```

Comment:
There's an ImmutableHashSet<T>.Builder you can use which is probably a bit more efficient than the conversion at the end.